mesh/mesh.go

```diff
@@ -255,6 +304,48 @@ func (msh *Mesh) setProcessedLayer(layerID types.LayerID) {
 	}
 }
 
+func (msh *Mesh) revertMaybe(ctx context.Context, logger log.Log, newVerified types.LayerID) error {
+	minUpdated := msh.getMinUpdatedLayer()
+	if minUpdated == (types.LayerID{}) {
+		// no contextual validity update since the last ProcessLayer() call
+		return nil
+	}
+	msh.resetMinUpdatedLayer(minUpdated)
+
+	var revertTo types.LayerID
+	for lid := minUpdated; !lid.After(newVerified); lid = lid.Add(1) {
+		block, _, err := msh.getBlockToApply(ctx, lid, newVerified)
+		if err != nil {
+			return err
+		}
+		hash := msh.GetLayerHash(lid)
+		if hash != types.EmptyLayerHash &&
+			hash != types.CalcBlocksHash32([]types.BlockID{block.ID()}, nil) {
```

Comment:
what if when we get hare output we set contextual validity of that block to true as well. and once minUpdated is not nil we know exactly which layer to apply. will it work?

Reply:
there are some concerns with this approach:
- minUpdated can be not nil but we don't necessarily have to revert. if block_A was applied, and then block_B later was also valid, we don't need to revert the state. just need to update the hash.
- in `GetHareConsensusOutput()` is mesh's only way to know if hare has terminated for that layer. we currently is relying on that to return an error to skip applying a layer when hare terminated out of order. if we solely rely on contextual validity, then we may end up applying an empty block for a late hare. for example, if hare terminates in this order:N->N+2->N+1. at the time N+2 terminated, we will apply an empty block for layer N+1
- in tortoise, there is also an assumption about the existence of contextual validity implies that tortoise has verified that layer. i am hesitant to break that assumption at this point.